scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala

```diff
@@ -0,0 +1,214 @@
+  implicit class AnnoyScioContext(val self: ScioContext) extends AnyVal {
+    /**
+      * Create a SideInput of `AnnoyIndex` from an [[AnnoyUri]] base path, to be used with
+      * [[com.spotify.scio.values.SCollection.withSideInputs SCollection.withSideInputs]]
+      */
+    def annoySideInput(metric: ScioAnnoyMetric, dim: Int, nTrees: Int, path: String)
+    : SideInput[AnnoyReader] = {
+      val uri = AnnoyUri(path, self.options)
+      val view = self.parallelize(Seq(uri)).applyInternal(View.asSingleton())
+      new AnnoySideInput(view, metric, dim, nTrees)
+    }
+  }
+
+  implicit class AnnoyPairSCollection(val self: SCollection[(Int, Array[Float])]) {
+    /**
+      * Write the key-value pairs of this SCollection as an Annoy file to a specific location.
+      *
+      * @return A singleton SCollection containing the [[AnnoyUri]] of the saved files
+      */
+    def asAnnoy(metric: ScioAnnoyMetric, dim: Int, nTrees: Int, path: String)
+    : SCollection[AnnoyUri] = {
+      val uri = AnnoyUri(path, self.context.options)
+      require(!uri.exists, s"Annoy URI ${uri.path} already exists")
+
+      self.transform { in =>
+        in.groupBy(_ => ())
+          .map { case(_, xs) =>
+            logger.info(s"Saving as Annoy: $uri")
+            val annoyIndex = new AnnoyWriter(metric, dim, nTrees)
+            val it = xs.iterator
+            while (it.hasNext) {
+              val kv = it.next()
+              annoyIndex.addItem(kv._1, kv._2)
+            }
+            uri.saveAndClose(annoyIndex)
+            uri
+          }
+      }
+    }
+
+    /**
+      * Write the key-value pairs of this SCollection as an Annoy file to a temporary location.
+      *
+      *   @return A Singleton SCollection containing the [[AnnoyUri]] of the saved files
+      */
+    def asAnnoy(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SCollection[AnnoyUri] = {
+      val uuid = UUID.randomUUID()
+      val path = self.context.options.getTempLocation + s"/annoy-build-$uuid"
+      this.asAnnoy(metric, dim, nTrees, path)
+    }
+
+    /**
+     * Write the key-value pairs of this SCollection as an Annoy file to a temporary location
+     * and load it as a SideInput[AnnoyReader]
+     * @return SideInput[AnnoyReader]
+     */
+    def asAnnoySideInput(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SideInput[AnnoyReader] = {
+      self.asAnnoy(metric, dim, nTrees).asAnnoySideInput(metric, dim, nTrees)
+    }
+  }
+
+  /**
+    * Enhanced version of [[com.spotify.scio.values.SCollection SCollection]] with Annoy methods
+    */
+  implicit class AnnoySCollection(val self: SCollection[AnnoyUri]) extends AnyVal {
+    def asAnnoySideInput(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SideInput[AnnoyReader] = {
+      val view = self.applyInternal(View.asSingleton())
+      new AnnoySideInput(view, metric, dim, nTrees)
+    }
+  }
+
+  private class AnnoySideInput(val view: PCollectionView[AnnoyUri],
```

Comment:
I assume there is no need to validate the constructor parameters or any other place earlier when we use it?

Reply:
Do you mean sanity checking that the values of metric, dim and nTrees fall in an appropriate range? I was relying on the user to be familiar enough with annoy to provide good values for that.

Reply:
I mean simple checks like dim larger than 0, there. There also was some hidden feature in annoy build, where if you set something to `-1` it will find optimal value etc.

Reply:
https://github.com/spotify/annoy/issues/216 and i think you could also set the number of trees to`-1` (https://github.com/spotify/annoy/commit/ef4051a8a15684222ce46208ff1a30b8b8a3ee35)